### PR TITLE
POC for Buffer Descriptor Analysis

### PR DESCRIPTION
**Pull request by jamestclin**
This draft shows an implementation of an analysis for generating Buffer Descriptors.
- Problem formulation:
    - Input: 
      - LinAlg IR converted from Triton IR by `--triton-to-linalg` conversion pass
      - Make dynamic strides/sizes from function argument to constant strides/sizes in arith.constant (task 3.2.1., this is done manually at the moment)
    - Output:
      - Buffer descriptor with constant length/steps/wraps, and dynamic base (base pointer address, a function of program_id)
- Proposed method in high-level:
    - Analysis is done by visiting all the `memref.copy/memref.tensor_store` in the function and recursively call visitor functions for operands until reaching `memref.reinterpretcast`. 
    - Buffer descriptor is propagated from `memref.reinterpretcast` to `memref.copy/memref.tensor_store` by recursion of visitor functions.
- Able to handle tasks 3.2.2. and 3.2.3. for triton tutorials:
    - add vector, softmax, and layernorm
    - matmul (works for various group sizes)
    - fused attention (only works for full attention that comes with constant for-loop tripcount. The original tutorial example is masked attention where the tripcount is dynamic to the program_id)
- Known limitation:
    - Memory block size masking/padding is skipped
    - Multiple load of a tensor in a single for loop iteration isn't supported
    - For-loop has to have constant tripcount to infer the BD for task 3.2.3.
    - Haven't looked into the case of nested for-loop

### CODE REVIEW

File: lib/Targets/BufferDescriptorAnalysis.cpp [hunk 1]
```diff
@@ -0,0 +1,391 @@
+//===- BufferDescriptorAnalysis.cpp -----------------------------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2023 Xilinx Inc.
```
Comment (stephenneuendorffer) [inline lib/Targets/BufferDescriptorAnalysis.cpp:7]: Advanced Micro Devices, Inc.

File: include/aie/Analysis/BufferDescriptorAnalysis.h [hunk 2]
```diff
@@ -0,0 +1,90 @@
+//===- BufferDescriptorAnalysis.h -----------------------------------------===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2023 Xilinx Inc.
+//
+//===----------------------------------------------------------------------===//
+
```
Comment (stephenneuendorffer) [inline include/aie/Analysis/BufferDescriptorAnalysis.h:10]: What does this do?  Is it really specific to your target, or is it more general?

File: lib/Targets/BufferDescriptorAnalysis.cpp (hunk 1 above)
Comment (jamestclin) [reply lib/Targets/BufferDescriptorAnalysis.cpp:7]: Will update the name here. Also, git checks show errors regarding clang formatting, and compilation warning. Will fix them as well.

File: include/aie/Analysis/BufferDescriptorAnalysis.h (hunk 2 above)
Comment (jamestclin) [reply include/aie/Analysis/BufferDescriptorAnalysis.h:10]: It's more to the target. Any suggestion on where this header file should live?

**Comment by stephenneuendorffer** [comment]
@jamestclin What's the status with this?

**Comment by jamestclin** [comment]
Close this PR for now, as the POC isn't in the current pursuit.

**Pull request closed without merging**